crates/forge_app/src/orch.rs

```diff
@@ -359,25 +359,56 @@ impl<S: AgentService> Orchestrator<S> {
-                        let sender = sender.clone();
-                        let agent_id = agent.id.clone();
-                        let model_id = model_id.clone();
-                        move |error: &anyhow::Error, duration: Duration| {
-                            tracing::error!(agent_id = %agent_id, error = %error, model=%model_id, "Retry Attempt");
-                            let retry_event = ChatResponse::RetryAttempt {
-                                cause: error.into(),
-                                duration,
-                            };
-                            let _ = sender.try_send(Ok(retry_event));
-                        }
-                    }),
-                )
-                .await?;
+            // Run the main chat request and compaction check in parallel
+            let main_request = crate::retry::retry_with_config(
+                &self.environment.retry_config,
+                || self.execute_chat_turn(&model_id, context.clone(), tool_supported, reasoning_supported),
+                self.sender.as_ref().map(|sender| {
+                    let sender = sender.clone();
+                    let agent_id = agent.id.clone();
+                    let model_id = model_id.clone();
+                    move |error: &anyhow::Error, duration: Duration| {
+                        tracing::error!(agent_id = %agent_id, error = %error, model=%model_id, "Retry Attempt");
+                        let retry_event = ChatResponse::RetryAttempt {
+                            cause: error.into(),
+                            duration,
+                        };
+                        let _ = sender.try_send(Ok(retry_event));
+                    }
+                }),
+            );
+
+            // Prepare compaction task that runs in parallel
+            let compaction_task = {
+                let agent_clone = agent.clone();
+                let context_clone = context.clone();
+                let services_clone = self.services.clone();
+                async move {
+                    // Estimate token count for compaction decision
+                    let estimated_tokens = context_clone.token_count();
+                    if agent_clone.should_compact(&context_clone, estimated_tokens) {
+                        info!(agent_id = %agent_clone.id, "Compaction needed, applying compaction in parallel");
+                        Compactor::new(services_clone)
+                            .compact(&agent_clone, context_clone.clone(), false)
+                            .await
+                            .map(Some)
+                    } else {
+                        debug!(agent_id = %agent_clone.id, "Compaction not needed");
+                        Ok(None)
+                    }
+                }
+            };
+
+            // Execute both operations in parallel
+            let (chat_result, compaction_result) = tokio::join!(main_request, compaction_task);
```

Comment:
```suggestion
            let (chat_result, compaction_result) = tokio::join!(main_request, compaction_task)?;
```


Better to propagate error right here.